Source/Dafny/AST/DafnyAst.cs

```diff
@@ -13,14 +13,92 @@
 using System.Diagnostics.Contracts;
 using System.Numerics;
 using System.Linq;
-using Microsoft.Boogie;
 using System.Diagnostics;
 using System.Threading;
+using Microsoft.Boogie;
 
 namespace Microsoft.Dafny {
   [System.AttributeUsage(System.AttributeTargets.Field)]
   public class FilledInDuringResolutionAttribute : System.Attribute { }
 
+  public interface IToken : Microsoft.Boogie.IToken {
```

Comment:
What's a token?

A usage of the word token that I'm familiar with is to refer to individual words in a program, so the program "method Foo(x: int)" would have the tokens "method, Foo, (, x, :, int, )" (https://en.wikipedia.org/wiki/Lexical_analysis)

Is that also what you mean by token here? Or do you use token to say, an object that exists for every AST node, that contains the syntax related information of that AST node, such as start and end position, and leading and trailing trivia?

If the answer is the former, where there's a token for every word in the program, then I wonder why you'd want to store that information.

If the answer is the latter, then I wonder if we should rename token to "NodeSyntax"

Reply:
It's the former, but let me widen your definition of what a token _is_ and _has_.

First, a token _is_ not just a "word", it's a word only for toy grammars and projects. In practice, beyond having a `val` (the word itself) and a potential initial `kind`, a token _has_ a filename from which it was extracted, _has_ a position, _has_ a line, _has_ a column. In our case, tokens even _have_ `next`, which is a pointer to the next token, making it possible to traverse the token list if the first one is available.
Essentially, the last three contain concrete formatting information that is not strictly needed in the scanning phase to build the AST. So, there is a precedent into attaching metadata to tokens.

Second, usual definition define tokens and extract them by stripping out whitespace. This is why we also call the resulting trees after parsing "abstract syntax tree", because they don't contain any formatting information.
I think we identified that it is not desirable to remove formatting information from the syntax tree for many reasons (docstring, translation with comments, auto-formatter, warnings if indentation makes something ambiguous, clever refactoring...).
Hence I will gradually make our syntax trees "concrete", so that they still retain (after parsing phase, at least), concrete information about parsing.

I know two ways of doing this efficiently:
* [Sketch-n-Sketch](https://github.com/ravichugh/sketch-n-sketch) builds the parse tree without going through the lexing phase, so there is no need for the notion of "token", and all the trivia is stored directly as fields of the AST. However, they don't make the distinction between "leading trivia" and "trailing trivia", which is sometimes cumbersome when refactoring.
* In our case, the scanning phase is separate from the parser phase. Moreover, there are 1 more trivia than tokens in any program, and each trivia can be split between what could visibly be attached to the previous token and what could be long to the following token. Thus, it is natural to me to make them "owned" by tokens themselves immediately. This scales better than going back to the file and take substrings according to positions when building the AST, potentially missing out trivia.

With that said, if the AST starts to be aware of tokens it was parsed on, we could simply verify that all tokens are owned by an AST node to ensure we are not missing trivia. This means more robust development down the road.

For now, I am attaching only the first token of a declaration to the declaration itself, because it's the token that typically would have the comment preceding the declaration. This might change as we are discussing where to attach docstring.